### Sections

1. [Introduction](#introduction)
2. [Loading the dataset into a DataFrame](#loading-the-dataset-into-a-dataframe)
3. [Choosing questions for our analysis](#choosing-questions-for-our-analysis)
4. [Converting data to the correct dtype](#converting-data-to-the-correct-dtype)  
    4.1. [Introduction](#4-introduction)  
    4.2. [Converting the dtype of the "duration (seconds)" column](#converting-the-dtype-of-the-duration-(seconds)-column)  
    4.3. [Converting the dtype of the "latitude" column](#converting-the-dtype-of-the-latitude-column)  
    4.4. [Converting the dtype of the "datetime" column](#converting-the-dtype-of-the-datetime-column)  
    4.5. [Converting the dtype of the "date posted" column](#converting-the-dtype-of-the-date-posted-column)
5. [Handling missing data](#handling-missing-data)  
    5.1. [Introduction](#5-introduction)  
    5.2. [Handling missing data in the shape column](#handling-missing-data-in-the-shape-column)  
    5.3. [Handling missing data in the "state" and "country" columns](#handling-missing-data-in-the-country-and-state-columns)  
6. [Removing duplicate data](#removing-duplicate-data)
7. [Saving the processed dataset to a file](#saving-the-processed-dataset-to-a-file)

# 1. Introduction <a id='introduction'></a>

In this notebook, we will be working with a dataset from [www.kaggle.com](www.kaggle.com) - a website that contains many different datasets and which regularly hosts data science competitions. 

The dataset we will be working with contains data regarding UFO sightings - you can learn more about it [here](https://data.world/timothyrenner/ufo-sightings). The first step is to load the dataset into a dataframe from a url, which can be done with the `read_csv()` function in the Pandas library. The code cell below stores the url to the dataset file  in the variable `dataset_url`

In [9]:
dataset_url = "https://raw.githubusercontent.com/lazarskiopencourses/lazarskiopencourses.github.io/master/courses/introduction_to_data_analysis/datasets/ufo_sightings.csv"

# 2. Loading the dataset into a DataFrame <a id='loading-the-dataset-into-a-dataframe'></a>

As you may have noticed, the dataset filename ends in ".csv". The "csv" stands for **c**omma-**s**eparated **v**alues. Csv files are commonly used to store tabular data, where the values for each column are separated by commas, and each row is on its own line.

For example, a csv file might look like this:

`name, breed, age`  
`Coco, Poodle, 4`  
`Teddy, Poodle, 2`  
`Nala, Husky, 6`  

The first row contains the names of the columns, whereas the subsequent rows contain the data. The above example is a small dataset consisting of 3 columns and 3 rows of data (4 rows total, but one row contains column names). The dataset we will be working with in this notebook is much bigger and contains more columns (11 columns), as well as many more rows (80332 rows).

The Pandas library has a function for loading a csv file into a `DataFrame` object. So first lets import the pandas library.

In [10]:
import pandas as pd

Now we can pass in the file path as an argument to the `read_csv()` function, which will return a `DataFrame` object containing the data in the csv file.

In [11]:
df = pd.read_csv(dataset_url)
df

/tmp/ipykernel_13024/602548147.py:1: DtypeWarning: Columns (5,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(dataset_url)


datetime                  city state country     shape  \
0      10/10/1949 20:30            san marcos    tx      us  cylinder   
1      10/10/1949 21:00          lackland afb    tx     NaN     light   
2      10/10/1955 17:00  chester (uk/england)   NaN      gb    circle   
3      10/10/1956 21:00                  edna    tx      us    circle   
4      10/10/1960 20:00               kaneohe    hi      us     light   
...                 ...                   ...   ...     ...       ...   
80327    9/9/2013 21:15             nashville    tn      us     light   
80328    9/9/2013 22:00                 boise    id      us    circle   
80329    9/9/2013 22:00                  napa    ca      us     other   
80330    9/9/2013 22:20                vienna    va      us    circle   
80331    9/9/2013 23:00                edmond    ok      us     cigar   

      duration (seconds) duration (hours/min)  \
0                   2700           45 minutes   
1                   7200              1-2 hrs   
2                     20           20 seconds   
3                     20             1/2 hour   
4                    900           15 minutes   
...                  ...                  ...   
80327              600.0           10 minutes   
80328             1200.0           20 minutes   
80329             1200.0                 hour   
80330                5.0            5 seconds   
80331             1020.0           17 minutes   

                                                comments date posted  \
0      This event took place in early fall around 194...   4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...   1/21/2008   
3      My older brother and twin sister were leaving ...   1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004   
...                                                  ...         ...   
80327  Round from the distance/slowly changing colors...   9/30/2013   
80328  Boise&#44 ID&#44 spherical&#44 20 min&#44 10 r...   9/30/2013   
80329                                       Napa UFO&#44   9/30/2013   
80330  Saw a five gold lit cicular craft moving fastl...   9/30/2013   
80331  2 witnesses 2  miles apart&#44 Red &amp; White...   9/30/2013   

         latitude  longitude   
0      29.8830556  -97.941111  
1        29.38421  -98.581082  
2            53.2   -2.916667  
3      28.9783333  -96.645833  
4      21.4180556 -157.803611  
...           ...         ...  
80327   36.165833  -86.784444  
80328   43.613611 -116.202500  
80329   38.297222 -122.284444  
80330   38.901111  -77.265556  
80331   35.652778  -97.477778  

[80332 rows x 11 columns]

At the bottom of the output in the previous cell we can see that the `DataFrame` has 80332 rows and 11 columns. We can also see that there is some warning displayed (in red) at the top of the output. The warning is about mixed types in columns 5 and 9 - we'll deal with that later, when we will be preparing our data for analysis.

The first thing we might want to do is to briefly take a look at the output above and familiarize ourselves with the dataset a bit. Here are some observations about the dataset that we can make, based on the `DataFrame` preview above:
1. Each row contains data related to a UFO sighting event.
2. Some columns contain quantitative data (duration, longitude, latitude, date)
3. Some columns contain qualitative data (city, state, country, shape, comments)
4. There are two "duration" columns which essentially contain the same data - just in a different format.
5. We can see that in the `state` and `country` columns, the following value appears: `NaN`. This is the default representation of missing values in Pandas (more will be said about this later).

# 3. Choosing questions for our analysis <a id='choosing-questions-for-our-analysis'></a>

Ok, so we have some sense of the kind of data this dataset contains. Now we can ask ourselves what kind of questions can we answer with this dataset. The questions we choose will in part shape our data preparation process.

For this dataset, let's say we want to answer the following questions:

1. What is the most commonly reported UFO shape?
2. What is the average duration of a UFO sighting?
3. When do UFO sightings occur?
4. Where do UFO sightings occur?

# 4. Converting data to the correct dtype <a id='converting-data-to-the-correct-dtype'></a>

## 4.1. Introduction <a id='4-introduction'></a>

When preparing data for analysis, it is good to start by making sure that the data in each column is of the correct `dtype`. For example, if we have a column containing numerical data, we want to store it as a numerical `dtype` (such as `float64` for example), because otherwise we might not be able to perform certain operations on the data in that column (such as arithmetic operations). 

Before we proceed to checking (and potentially correcting) the `dtype` of each column in our dataframe, it is worth knowing two things:

1. By default, columns with strings are assigned the `object` dtype in Pandas. The `object` dtype actually corresponds to any Python object - a list, a dictionary, a string, an integer etc. This means that a column with the `object` dtype can store any Python object. It is also important to know that the `object` dtype is the default `dtype` in Pandas. This means that columns with data of mixed types, or data whose `dtype` cannot be inferred by Pandas, will be assigned the `object` dtype by default. 


2. Pandas (and NumPy) have a special `dtype` for dates and times, since this kind of data is quite common. The `datetime64` dtype is used for storing both the date and time (as the name suggests).

Ok, so if we want to check the data types of our columns, we can use the `DataFrame.dtypes` attribute:

In [3]:
df.dtypes

datetime                 object
city                     object
state                    object
country                  object
shape                    object
duration (seconds)       object
duration (hours/min)     object
comments                 object
date posted              object
latitude                 object
longitude               float64
dtype: object

We can see that almost all of the columns are of the default `object` dtype. For columns such as "country" or "comments" this is ok, since those columns only contain strings. However, columns "duration (seconds)" and "latitude" contain numerical data. Furthermore, the column "datetime" contains dates and times, and the column "date posted" contains just dates.

Therefore the following columns should be converted to the correct dtypes:

    
|Column name          | Current dtype| Target dtype|
|:--------------------|:-------------|:------------|
|duration (seconds)   |object        | int64       |
|latitude             |object        | float64     |
|datetime             |object        | datetime64  |
|date posted          |object        | datetime64  |

## 4.2. Converting the dtype of the "duration (seconds)" column <a id='converting-the-dtype-of-the-duration-(seconds)-column'></a>

We can convert columns to different data types by using the `Series.astype()` method. Lets start with the "duration (seconds)" column.

**Note:** running the code cell below will result in an error - this is supposed to happen.

In [4]:
df["duration (seconds)"] = df["duration (seconds)"].astype("int64")

ValueError: invalid literal for int() with base 10: '1.5'

Ok, something went wrong because we see a long error message. If we look at the very bottom of the error message, we can see: `ValueError: invalid literal for int() with base 10: '1.5'`. Pandas is trying to tell us that it cannot convert the string `'1.5'` to an integer. Since we asked Pandas to convert all of the data in the "seconds (duration)" column to integers, this means that somewhere in that column there is a string with the value `'1.5'` and it is causing this error. 

We can also suspect that the problem is connected with the fact that the string contains a decimal number, whereas integers do not have a fractional part. Nevertheless, we have discovered that our initial assumption about the data is incorrect. The "duration (seconds)" column contains fractions of seconds and therefore `int64` is not a good `dtype` to convert to, since we will lose some of the precision in our data. In this case, let's try to convert the column `dtype` to `float64` and see if we are successful this time:

In [5]:
df["duration (seconds)"] = df["duration (seconds)"].astype("float64")

ValueError: could not convert string to float: '2`'

And we got another error message. This time Pandas is telling us that it cannot convert the string ```'2`'``` to a `float`. We can see that there is something something strange with this string - it contains the following character: ``` ` ```. Its unclear what the ``` ` ``` is supposed to mean. It is most likely a mistake and it should not be there. We can also suspect that it is also the cause of this error, as Pandas does not know how to deal with this character when converting the string to a `float`.

Let's fix this by replacing the value ```'2`'``` with `'2'`. We can use do this with the `Series.replace()` method.

In [6]:
df["duration (seconds)"] = df["duration (seconds)"].replace("2`", "2")

Now lets try to convert the dtype of this column to `float64` again:

In [7]:
df["duration (seconds)"] = df["duration (seconds)"].astype("float64")

ValueError: could not convert string to float: '8`'

Ok, now we we see a similar error message, but this time concerning the string ```'8`'```. It seems that the ``` ` ``` character appears multiple times throughout this column. Let's fix this by removing all occurrences of this character in the "duration (seconds)" column. One way to do this is to replace this character with an empty string. We can do this with the `Series.str.replace()` method, which will match substrings as well, not just entire the string.

However, before we do this, we should convert the dtype of the "duration (seconds)" column to string. This is because the current dtype is `object`, which means that this column can contain any Python object - including a mix of different types of objects. The "duration (seconds)" column most likely contains a mix of `int`, `float` and `str`. If we use the `Series.str.replace()` function right now, it may cause unintended consequences, as this function expects all the values in this column to be strings. Therefore, it is best to first convert all the values in the column to strings.

In [8]:
df["duration (seconds)"] = df["duration (seconds)"].astype("str")

Ok, now let's replace all occurrences of ``` ` ``` with an empty string:

In [9]:
df["duration (seconds)"] = df["duration (seconds)"].str.replace("`", "")

Ok, lets try converting the dtype once again:

In [10]:
df["duration (seconds)"] = df["duration (seconds)"].astype("float64")

Hurray, no error message this time! So why did we do this? Well, since our column now has a numerical `dtype`, we can easily do all sorts of useful operations on it, which would not have been possible if the `dtype` was `object`. Here are some examples:

In [11]:
print("min:", df["duration (seconds)"].min())
print("max:", df["duration (seconds)"].max())
print("mean:", df["duration (seconds)"].mean())
print("median:", df["duration (seconds)"].median())

min: 0.001
max: 97836000.0
mean: 9016.889016344669
median: 180.0


**Note:** in the code cell above, we pass in two arguments into the `print()` function. You can pass in an arbitrary number of arguments to the print function, which by default prints all of those arguments separated by white space `' '`. Therefore, the two lines below will print the same thing:

In [12]:
print("a" + " " + "b")
print("a", "b")

a b
a b


## 4.3. Converting the dtype of the "latitude" column <a id='converting-the-dtype-of-the-latitude-column'></a>

Let's correct the dtype of the "latitude" column next:

In [13]:
df["latitude"].astype("float64")

ValueError: could not convert string to float: '33q.200088'

The latitude should just be a number, but there seems to be a letter there, as indicated by the error message. Let's remove all occurrences of this letter (just in case it shows up multiple times):

In [14]:
df["latitude"] = df["latitude"].astype("str")
df["latitude"] = df["latitude"].str.replace("q", "")

Now let's try to convert the dtype again:

In [15]:
df["latitude"] = df["latitude"].astype("float64")
df["latitude"].dtype

dtype('float64')

We have successfully converted the `dtype`, let's move on to the next column...

## 4.4. Converting the dtype of the "datetime" column <a id='converting-the-dtype-of-the-datetime-column'></a>

Now let's correct the dtype of the "datetime" column. When converting a column containing strings with dates and times, we should first identify the format in which the date and time is represented. For example, here are three different formats for representing the date and time:

1. `05-SEP-17 18:37:21.4931`
2. `12:35 24/02/2017`
3. `08-23-1978 7:42PM`

Pandas also has the ability to infer the format of a datetime string, however, before we use this functionality, we should know what the format is ourselves. This will enable us to verify that Pandas has correctly inferred the format of the date and time.

Let's preview the "datetime" column:

In [16]:
df["datetime"]

0        10/10/1949 20:30
1        10/10/1949 21:00
2        10/10/1955 17:00
3        10/10/1956 21:00
4        10/10/1960 20:00
               ...       
80327      9/9/2013 21:15
80328      9/9/2013 22:00
80329      9/9/2013 22:00
80330      9/9/2013 22:20
80331      9/9/2013 23:00
Name: datetime, Length: 80332, dtype: object

Based on the data displayed above, we cannot determine which numbers represent the day and which numbers represent the month - we only see the numbers `9` and `10`. So first let's try to determine whether the first number represents the day or the month. There are many ways to do this. Perhaps one simple way is to get a random sample of dates and times from the "datetime" column and hope that one of them will contain a number bigger than 12 on the first or second position. 

Let's try this first. We can use the `Series.sample()` method to get a random sample of values from a `Series`:

In [17]:
df["datetime"].sample(10)

45690      5/5/2004 19:00
44250     5/25/2002 22:04
55273      7/1/1987 03:00
30346     3/10/2012 23:00
7721     11/10/2005 21:00
9208     11/14/2010 15:32
15322     11/6/2007 05:10
26752     2/15/2008 24:00
54701     7/10/2002 23:56
41810      5/1/2014 23:00
Name: datetime, dtype: object

Since the `Series.sample()` method returns a random sample, the values you see will almost certainly be different each time you run the above code cell. If you do not see any number above 12 on the second position, go ahead and run the code cell above again.

If you have successfully obtained a date with a day larger than 12, you can see that the first number represents the month and the second number represents the day. So our datetime string has the following format: `month/day/year hour:minute`

It is also important to notice that single digit month numbers do not have a preceding eg. `3/24/2001`, whereas single digit hours do, eg. `07:35`.

Since we now know what the format of the datetime string is, we can convert the dtype of the "datetime" column using the `pd.to_datetime()` function. We will also set the `infer_datetime_format` argument to `True`, to see if Pandas can do some work for us.

In [18]:
pd.to_datetime(df["datetime"], infer_datetime_format=True)

ParserError: hour must be in 0..23: 10/11/2006 24:00

Ok, so we got an error message from Pandas. If we look at the bottom of the error message, it seems to be suggesting that the hours must be between `0` and `23`. We can also see that the error message contains some datetime string (presumably from the "datetime" column) which has the number `24` as the hour. 

Let's replace all those `24` hours with `00`, however, we must be careful not to replace numbers corresponding to the 24th day. One clever way to do this is to replace all the `24:` substrings with `00:`. By adding the colon `:` we ensure that only the hours will be replaced, since the days, months and years use the forward slash `/` as a separator.

In [19]:
df["datetime"] = df["datetime"].astype("str")
df["datetime"] = df["datetime"].str.replace("24:", "00:")

Ok, lets try to convert the dtype of the

In [20]:
pd.to_datetime(df["datetime"], infer_datetime_format=True)

0       1949-10-10 20:30:00
1       1949-10-10 21:00:00
2       1955-10-10 17:00:00
3       1956-10-10 21:00:00
4       1960-10-10 20:00:00
                ...        
80327   2013-09-09 21:15:00
80328   2013-09-09 22:00:00
80329   2013-09-09 22:00:00
80330   2013-09-09 22:20:00
80331   2013-09-09 23:00:00
Name: datetime, Length: 80332, dtype: datetime64[ns]

It seems that Pandas was able to infer the datetime format on its own. If it would not have been able to, we would have to manually pass in the datetime format as the `format` argument in the `pd.to_datetime()` function. The `format` argument should be a string containing a specific sequence of characters, which corresponds to our datetime format - you can learn more about this on the [official documentation page](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html).

**Note:** The code cell above returns a new `Series` object, but we did not assign it to `df["datetime"]`, so let's do that now.

In [21]:
df["datetime"] = pd.to_datetime(df["datetime"], infer_datetime_format=True)

Since we have successfully converted the dtype of our "datetime" column, we can now easily do various useful operations. For example, we can extract the second/minute/hour/day/month/year out of each datetime. The code cell below extracts the hour:

In [22]:
df["datetime"].dt.hour

0        20
1        21
2        17
3        21
4        20
         ..
80327    21
80328    22
80329    22
80330    22
80331    23
Name: datetime, Length: 80332, dtype: int64

We can also use part of the datetime to create masks and filter the dataframe:

In [23]:
mask = df["datetime"].dt.month == 1
df[mask]["datetime"]

7407    1947-01-10 20:00:00
7408    1959-01-10 18:30:00
7409    1964-01-10 20:00:00
7410    1968-01-10 23:00:00
7411    1971-01-10 20:00:00
                ...        
25495   2014-01-09 18:44:00
25496   2014-01-09 19:15:00
25497   2014-01-09 19:15:00
25498   2014-01-09 21:20:00
25499   2014-01-09 21:50:00
Name: datetime, Length: 5689, dtype: datetime64[ns]

In [24]:
mask = df["datetime"].dt.year > 1990
df[mask]["datetime"]

49      1991-10-10 22:00:00
50      1991-10-10 22:00:00
51      1992-10-10 17:00:00
52      1992-10-10 18:00:00
53      1992-10-10 20:15:00
                ...        
80327   2013-09-09 21:15:00
80328   2013-09-09 22:00:00
80329   2013-09-09 22:00:00
80330   2013-09-09 22:20:00
80331   2013-09-09 23:00:00
Name: datetime, Length: 73781, dtype: datetime64[ns]

There are also many other useful things we can do, which you can read about on the [official documentation page](https://pandas.pydata.org/docs/user_guide/timeseries.html). As always, it is not important to memorize all of the operations that can be performed on datetime objects in Pandas. It is only important to know that there are a lot of different operations and that if we are trying to do something with datetime objects, it might be worth reading the official documentation or other sources, to try and see if there is a function, which allows us to accomplish our task easily.

## 4.5. Converting the dtype of the "date posted" column <a id='converting-the-dtype-of-the-date-posted-column'></a>

Ok, let's try to convert the dtype of the "date posted" column to `datetime` and let Pandas handle the datetime format:

In [25]:
df["date posted"] = pd.to_datetime(df["date posted"], infer_datetime_format=True)

No error message? Well sometimes the stars align. Let's preview the column to make sure everything is ok - just in case:

In [26]:
df["date posted"]

0       2004-04-27
1       2005-12-16
2       2008-01-21
3       2004-01-17
4       2004-01-22
           ...    
80327   2013-09-30
80328   2013-09-30
80329   2013-09-30
80330   2013-09-30
80331   2013-09-30
Name: date posted, Length: 80332, dtype: datetime64[ns]

Everything looks good! Let's move on to handling missing data.

# 5. Handling missing data <a id='handling-missing-data'></a>

## 5.1 Introduction <a id='5-introduction'></a>

When preparing data for analysis, it is good to start by looking for missing values. In general, there are two things to look for specifically:

**1. Missing values that affect our analysis somehow** - these need to be dealt with either by excluding rows with missing values from our analysis, or by doing imputation (replacing missing values with estimates).

**2. Missing values that have not been registered properly in Pandas** - the Pandas library has specific types of objects dedicated to representing missing values. Sometimes a column might contain strings such as "unknown", "N/A", "?", which often mean that the data is missing. However, Pandas considers these values to be regular strings instead of a missing data. In this case, it is useful to convert these types of strings to missing values, so that we can easily exclude them from our analyses whenever we need to.

One way to check which columns have missing values is to use the `DataFrame.isna()` method combined with the `DataFrame.sum()` method. 

First let's cover the `DataFrame.isna()` method, which checks for values that are not available (NA). This method returns a new `DataFrame` with each value replaced either by `True` (if it is missing) or `False` (if it is available)

In [27]:
df.isna()

datetime   city  state  country  shape  duration (seconds)  \
0         False  False  False    False  False               False   
1         False  False  False     True  False               False   
2         False  False   True    False  False               False   
3         False  False  False    False  False               False   
4         False  False  False    False  False               False   
...         ...    ...    ...      ...    ...                 ...   
80327     False  False  False    False  False               False   
80328     False  False  False    False  False               False   
80329     False  False  False    False  False               False   
80330     False  False  False    False  False               False   
80331     False  False  False    False  False               False   

       duration (hours/min)  comments  date posted  latitude  longitude   
0                     False     False        False     False       False  
1                     False     False        False     False       False  
2                     False     False        False     False       False  
3                     False     False        False     False       False  
4                     False     False        False     False       False  
...                     ...       ...          ...       ...         ...  
80327                 False     False        False     False       False  
80328                 False     False        False     False       False  
80329                 False     False        False     False       False  
80330                 False     False        False     False       False  
80331                 False     False        False     False       False  

[80332 rows x 11 columns]

Just as a side note, sometimes you may see people use the `DataFrame.isnull()` method. The `.isna()` and `.isnull()` methods are identical and you can use either one.

In [28]:
df.isnull()

datetime   city  state  country  shape  duration (seconds)  \
0         False  False  False    False  False               False   
1         False  False  False     True  False               False   
2         False  False   True    False  False               False   
3         False  False  False    False  False               False   
4         False  False  False    False  False               False   
...         ...    ...    ...      ...    ...                 ...   
80327     False  False  False    False  False               False   
80328     False  False  False    False  False               False   
80329     False  False  False    False  False               False   
80330     False  False  False    False  False               False   
80331     False  False  False    False  False               False   

       duration (hours/min)  comments  date posted  latitude  longitude   
0                     False     False        False     False       False  
1                     False     False        False     False       False  
2                     False     False        False     False       False  
3                     False     False        False     False       False  
4                     False     False        False     False       False  
...                     ...       ...          ...       ...         ...  
80327                 False     False        False     False       False  
80328                 False     False        False     False       False  
80329                 False     False        False     False       False  
80330                 False     False        False     False       False  
80331                 False     False        False     False       False  

[80332 rows x 11 columns]

Once we have a `DataFrame` containing `True` and `False` values, we can use the `DataFrame.sum()` method to count the number of `True` values. This is because `True` and `False` are usually associated with the numbers 1 and 0, where `True == 1` and `False == 0`.

In [29]:
True == 1

True

In [30]:
False == 0

True

If we call the `Series.sum()` method, it will return the sum of all values in a `Series`. If we call the `DataFrame.sum()` method, it will return the sum of each `Series`/column separately (stored in a `Series`, where the indices are column names and the values are the sums). The values `True` and `False` are treated as 1's and 0's, and therefore the sum in this case is simply the number of `True` values in a given series.

Therefore, if we call `DataFrame.isna()` we get a dataframe with booleans, where each `True` value corresponds to a missing value. If we call the `.sum()` method on such a dataframe, we will receive a Series containing the number of missing values for each column in our original dataframe.

In [31]:
df.isna().sum()

datetime                   0
city                       0
state                   5797
country                 9670
shape                   1932
duration (seconds)         0
duration (hours/min)       0
comments                  15
date posted                0
latitude                   0
longitude                  0
dtype: int64

As we can see, there are 4 columns with missing values. However, we might want to display the number of missing values as a percentage of the total number of possible values. In this case, we can first check the number of rows in the dataframe using the `len()` function:

In [32]:
len(df)

80332

Next we can divide the series with missing values by the number of rows in the dataframe:

In [33]:
df.isna().sum() / len(df)

datetime                0.000000
city                    0.000000
state                   0.072163
country                 0.120375
shape                   0.024050
duration (seconds)      0.000000
duration (hours/min)    0.000000
comments                0.000187
date posted             0.000000
latitude                0.000000
longitude               0.000000
dtype: float64

That works, but there's a lot of number and things are a bit difficult to read. Let's rewrite the previous code and also apply a mask to only see the columns with missing values.

In [34]:
s_missing = df.isna().sum()
s_missing = s_missing / len(df)
s_missing[s_missing > 0]

state       0.072163
country     0.120375
shape       0.024050
comments    0.000187
dtype: float64

Ok, that's better. We can now see more clearly which columns have missing values.

If we wanted to, we could also use the `Series.round()` method to round the values to 2 decimal places and improve readability even further. However, in this case, we have to remember that the percentage of missing values in the "comments" column will be displayed as 0 due to rounding. 

In [35]:
s_missing[s_missing > 0].round(2)

state       0.07
country     0.12
shape       0.02
comments    0.00
dtype: float64

It is also worth knowing that there is a `.notna()` method which does the opposite of the `.isna()` method. That is, it replaces each value with `True` if the value is not missing - otherwise it replaces is with `False`

In [36]:
df.notna()

datetime  city  state  country  shape  duration (seconds)  \
0          True  True   True     True   True                True   
1          True  True   True    False   True                True   
2          True  True  False     True   True                True   
3          True  True   True     True   True                True   
4          True  True   True     True   True                True   
...         ...   ...    ...      ...    ...                 ...   
80327      True  True   True     True   True                True   
80328      True  True   True     True   True                True   
80329      True  True   True     True   True                True   
80330      True  True   True     True   True                True   
80331      True  True   True     True   True                True   

       duration (hours/min)  comments  date posted  latitude  longitude   
0                      True      True         True      True        True  
1                      True      True         True      True        True  
2                      True      True         True      True        True  
3                      True      True         True      True        True  
4                      True      True         True      True        True  
...                     ...       ...          ...       ...         ...  
80327                  True      True         True      True        True  
80328                  True      True         True      True        True  
80329                  True      True         True      True        True  
80330                  True      True         True      True        True  
80331                  True      True         True      True        True  

[80332 rows x 11 columns]

Therefore, we could also obtain the series with missing value percentages by first using the `.notna()` method in conjunction with `sum()` and `len(df)` to obtain the percentage of non-missing values. Next we can subtract this percentage from 1 to obtain the percentage of missing values.

In [37]:
s_missing_2 = df.notna().sum() / len(df)
s_missing_2 = 1 - s_missing_2
s_missing_2 = s_missing_2[s_missing_2 > 0].round(2)
s_missing_2

state       0.07
country     0.12
shape       0.02
comments    0.00
dtype: float64

## 5.2 Handling missing data in the "shape" column <a id='handling-missing-data-in-the-shape-column'></a>

Ok, so the first question we want to answer in our analysis is "what is the most commonly reported UFO shape?". From this question we can deduce that we will be using the "shape" column. Most likely we will be calculating the mode of the data in the "shape" column (the mode is the most frequently occurring value).

Based on `s_missing`, we know that roughly 2% of the data is missing in the "shape" column. Since we're just going to be calculating the mode, and the vast majority of the data is there, this will not have a significant impact on our findings. Additionally, when presenting the findings of our analysis, we can convey the information about missing data and discuss it's relevancy.

However, we should check whether there are any missing values that were not properly registered. To do this, we can use the `Series.unique()` method, which returns an array of all the unique values in a series. 

In [38]:
df["shape"].unique()

array(['cylinder', 'light', 'circle', 'sphere', 'disk', 'fireball',
       'unknown', 'oval', 'other', 'cigar', 'rectangle', 'chevron',
       'triangle', 'formation', nan, 'delta', 'changing', 'egg',
       'diamond', 'flash', 'teardrop', 'cone', 'cross', 'pyramid',
       'round', 'crescent', 'flare', 'hexagon', 'dome', 'changed'],
      dtype=object)

Firstly, it is worth pointing out that one of the values in the "shape" column is: `nan`. Notice that there are no quotation marks around the `nan` - this is because it is not a string. The `nan` value is the default representation for missing data in Pandas. As we have already figured out previously, roughly 2% of the values in the "shape" column are `nan`. 

The `nan` object used to represent missing data is actually an object from the numpy library (`numpy.nan`). There are also other objects, which can used to represent missing data, such as the built-in Python object `None`. Multiple objects representing missing data is not ideal as it can get a little confusing. However, the most important thing to remember is that `nan` (`numpy.nan`), `None` and `<NA>` (`pandas.NA`) are meant to represent the same thing - missing data - even though they are different objects in Python.

The second thing worth pointing out is that there is an `'unknown'` value. Here it is not clear whether the  string `'unknown'` refers to the fact that a person could not identify the shape of the UFO, or whether the person did not provide any data about the shape (for whatever reason). In the latter case, this value should be counted as a missing value because the data is unavailable.

We could try to dig deeper into the origins of this dataset and see whether there is any information on the internet that could help us figure out which is it. If we do not manage to find an answer, we may wish to include this information during the presentation of our analysis.

For the purpose of this course, we will assume that the string `'unknown'` should indeed be counted as a missing value. This will allow for a demonstration of how one can add missing values to a dataframe. To do so, we can use the `Series.replace()` method, which enables us to replace one value with another.

In [40]:
print("Number of missing values before: " + str(df["shape"].isna().sum()))

df["shape"] = df["shape"].replace("unknown", None)

print("Number of missing values after: " + str(df["shape"].isna().sum()))

Number of missing values before: 1932
Number of missing values after: 7516


## 5.3 Handling missing data in the "country" and "state" columns <a id='handling-missing-data-in-the-country-and-state-columns'></a>

Since in our analysis we will be answering the question "where are UFO sightings most common?", we should examine the missing data in the "country" and "state" columns. Let's first take a look at those columns:

In [41]:
cols = ["state", "country"]
df[cols]

state country
0        tx      us
1        tx     NaN
2       NaN      gb
3        tx      us
4        hi      us
...     ...     ...
80327    tn      us
80328    id      us
80329    ca      us
80330    va      us
80331    ok      us

[80332 rows x 2 columns]

Ok, the first thing we might observe is that there is a row with a missing value in the "country" column, however, the corresponding "state" value is not missing. This means we can fill in the missing country value based on the state value.  

The second thing we might observe is that there is a row with a missing value in the "state" column, however, the corresponding "country" value is not missing. In this case, we might want to consider that the "state" column may contain "valid" missing values. More specifically, some values can be missing because the "state" column is not applicable to all countries. For example, Great Britain (presumably labeled with the identifier "gb") is not divided into states the same way as the US, so maybe this is why the value is missing. Perhaps in this dataset only US countries have some value in the "state" column. Let's see if this is true:

In [42]:
mask = df["state"].notna()
df[mask]["country"].unique()

array(['us', nan, 'ca', 'au', 'gb'], dtype=object)

It appears our initial guess was wrong - in this dataset there are four countries that have values in the "state" column (the `nan` represents a missing value, meaning some rows have a missing value in the "country" column, but a non-missing value in the "state" column). Let's also check what countries there are in general, just to get a better sense of the dataset as a whole: 

In [43]:
df["country"].unique()

array(['us', nan, 'gb', 'ca', 'au', 'de'], dtype=object)

It appears that there are 5 different countries and only Germany (presumably labeled with "de") does not have any values in the "state" column associated with it. Ok, let's check how many "state" values are missing, when we exclude Germany from the count, since we will assume that the missing "state" values there are valid. First let's create a mask that filters out Germany from the dataset:

In [44]:
mask_no_germany = df["country"] != "de"

Now lets create a mask to only includes rows with missing values in the "state" column:

In [45]:
mask_missing_state = df["state"].isna()

If we want to combine those masks, we might try to first apply one mask and then apply the other mask. However, this is not a recommended way of doing things, since once we apply the first mask, the number of rows in the dataframe will be smaller than the number of values in the second mask. You may remember that a mask is actually a boolean array, where each value corresponds to a row. A value of `True` means that the row is filtered out, whereas `False` means that the row is kept. Therefore a mismatch between the number of rows in the dataframe and the number of values in the mask can cause problems. Let's verify that this mismatch exists using the `len()` function:

In [46]:
print("num rows in df with no mask:", len(df))
print("num values in 1st mask:", len(mask_no_germany))
print("num rows in df with 1st mask:", len(df[mask_no_germany]))
print("num values in 2nd mask:", len(mask_missing_state))

num rows in df with no mask: 80332
num values in 1st mask: 80332
num rows in df with 1st mask: 80227
num values in 2nd mask: 80332


As we can see, the number of rows is indeed different after the first mask has been applied. As a side note, we have also discovered that a very small number of rows contain "de" in the "country" column.

If we want to apply two masks together, it is best to either:

1. first create and apply the first mask. Then create the second mask based on the new filtered dataframe, and then apply the second mask.

2. combine both masks together into one mask using the `&` operator.

Let's choose the second option. The `&` operator works analogously to the `and` operator, in that it performs an "and" operation on each pair of corresponding values in the two masks, and returns a new mask, which contains `True` only in locations where both the first and second masks are also `True`. Let's apply the combined mask and preview the relevant columns in the dataframe:

In [47]:
df[mask_no_germany & mask_missing_state][cols]

state country
2       NaN      gb
6       NaN      gb
18      NaN     NaN
20      NaN      gb
24      NaN      gb
...     ...     ...
80214   NaN      gb
80217   NaN      au
80234   NaN      gb
80254   NaN      gb
80322   NaN     NaN

[5692 rows x 2 columns]

As we can see at the bottom of the output above, there are 5692 rows with missing values in the "state" column and values other than "de" in the "country" column. These are missing values that should presumably not be missing. Let's see what percentage they constitute of the entire dataset:

In [48]:
len(df[mask_no_germany & mask_missing_state]) / len(df)

0.07085594781656127

It appears that 7.09% of the "state" values are missing. At this point we can ask ourselves a few different questions like: what is the meaning of the state column for countries other than the US and are some of the missing state values for those countries valid (due to inapplicability)? Another question we can ask ourselves is whether or not these missing values are even relevant to our analysis? Perhaps we will not be analyzing the location of UFO sightings based on states, but only based on the countries and cities. In this case, the missing values in the "state" column do not affect our analysis, since we will not be using the data in that column at all. If that would be the case, we would not have do anything with those missing values and we could just proceed to the next step in our data preparation process.

For the purpose of this course, let's say we want to show the frequency of UFO sightings in different states but only for the US. Let's check whether there are any missing state values for the US:

In [49]:
mask = (df["country"] == "us") & df["state"].isna()
len(df[mask])

0

It appears that there are no missing state values for the US. In that case, we can move on to handling the missing values in the "country" column, which definitely should be filled in if possible. Luckily, we can infer the country values based on the values in either the "city" column or the "state" column. Let's update the `cols` variable and preview the dataset once again, just to get a better sense of what we're working with:

In [50]:
cols = ["city", "state", "country"]
df[cols]

city state country
0                san marcos    tx      us
1              lackland afb    tx     NaN
2      chester (uk/england)   NaN      gb
3                      edna    tx      us
4                   kaneohe    hi      us
...                     ...   ...     ...
80327             nashville    tn      us
80328                 boise    id      us
80329                  napa    ca      us
80330                vienna    va      us
80331                edmond    ok      us

[80332 rows x 3 columns]

So we want to fill in missing country values based on the "city" or "state" column. Let's think about this: there are rows that have a missing value in the "country" column, but have some value in the "city" or "state" column. There might be some other rows with the same value in the "city" or "state" columns, that also have some non-missing value in the "country" column. Therefore, for each city and state, we could check if there is at least one row, which has a non-missing country value, and then use this value to overwrite all the missing country values. 

This seems like a good way to fill in some missing values, without referring to external data. However, there is something we should consider: what if two countries have a city with the same name or a state with the same identifier. We'll have to account for both of those things in our code, in order to ensure maximum accuracy of our data. Let's create a function that fills in the missing country values either based on the "city" column or the "state" column:

In [51]:
def fill_in_missing_countries(df, col_to_use):
    
    # check whether the argument `col_to_use` is either "city" or "state"
    if col_to_use not in ("city", "state"):
        raise ValueError('argument `col_to_use` must be either "city" or "state"')
    
    # for loop over each city/state
    # NOTE: we want to exclude missing values, therefore we use the `Series.dropna()`
    # method, which removes missing values from the Series
    for location in df[col_to_use].dropna().unique():

        # create a mask based on the location
        mask = df[col_to_use] == location
        
        # get all of the unique country values for the current location (excluding missing values)
        countries = df[mask]["country"].dropna().unique()
        
        # if there is exactly one non-missing country value, overwrite all country values
        # for that location (this will essentially just overwrite the missing state values)
        if len(countries) == 1:
            df.loc[mask, "country"] = countries[0] 

You may have noticed that the function above uses the reserved keyword `raise` and the built-in object `ValueError`. In Python, it is possible to "manually" force the Python interpreter to halt the execution of a program and display an error message. This is called raising an exception. An exception is just an error that occurs while a Python program is running. For example, the code cell bellow raises a `ValueError` - a type of exception:

In [52]:
print("start")
raise ValueError("custom error message")
print("end")

start


ValueError: custom error message

As we can see, the `print("end")` statement was not executed. There are many different types of errors in Python, many of which you have likely already seen: `NameError`, `TypeError`, `ValueError` etc. These can exceptions can be raised in our code whenever we want to, in order to prevent unwanted behavior.

For example, the `fill_in_missing_countries()` function accepts any value as the `col_to_use` argument. However, in reality, this function was created with only two columns in mind - the "city" and "state" columns. Passing in some other column name could result in unintended consequences, however, Python by itself wouldn't necessarily raise an exception. If we come back to this code in the future, we might forget this - "manually" raising an exception gives us additional protection from those unintended consequences. Similarly, if someone else was to use our function, they might accidentally use our function "incorrectly". This is why adding statements which raise exceptions to our code can be a good idea.

Ok, so we have defined a function that will fill in the missing country values. Let's check again what percentage of country values are missing:

In [53]:
mask = df["country"].isna()
len(df[mask]) / len(df)

0.1203754419160484

Ok, so around 12% of the country values are missing. Let's call the `fill_in_missing_countries()` function and fill in some of those values based on the "city" column:

**Note:** the code in the cell below can take up to 3 minutes to execute.

In [54]:
fill_in_missing_countries(df, "city")

Let's check what percentage of country values are missing now:

In [55]:
mask = df["country"].isna()
len(df[mask]) / len(df)

0.11347906189314345

We've managed to decrease the percentage of missing country values by a bit. Now let's call the `fill_in_missing_countries()` function again, but this time let's pass in "state" as the `col_to_use` argument:

In [56]:
fill_in_missing_countries(df, "state")
mask = df["country"].isna()
len(df[mask]) / len(df)

0.05734949957675646

This time we've filled in more missing country values, bringing down the total percentage of missing country values to about 5.7%. Let's move on to the next step in our data preparation process.

# 6. Removing duplicate data <a id='removing-duplicate-data'></a>

The second question we want to answer in our analysis pertains to the duration of UFO sightings. Therefore, we will be using the "duration" column. However, there are two columns that contain data regarding the duration of the UFO sighting. Let's remove one of those columns, since having both of those columns means we have unnecessary, duplicate data.

First let's take a look at those two columns and decide which one to remove.

In [57]:
col_names = ["duration (seconds)", "duration (hours/min)"]
df[col_names]

duration (seconds) duration (hours/min)
0                  2700.0           45 minutes
1                  7200.0              1-2 hrs
2                    20.0           20 seconds
3                    20.0             1/2 hour
4                   900.0           15 minutes
...                   ...                  ...
80327               600.0           10 minutes
80328              1200.0           20 minutes
80329              1200.0                 hour
80330                 5.0            5 seconds
80331              1020.0           17 minutes

[80332 rows x 2 columns]

**Note**: there is some discrepancy between the data in the "duration (seconds)" column and the "duration (hours/min)" column. For the purpose of this course, we will ignore it; however, in a real-life situation, it would be a good idea to figure out, which column contains the more accurate data (if possible); alternatively, one could measure the discrepancy between the data in the two columns and discuss the impact it has on the analysis.

Data pertaining to the duration of an event is in essence numerical. Therefore, the "duration (seconds)" column looks like a better candidate to keep at first glance, since it seems to contain only numbers and no text. The "duration hours/min" column contains both numbers and text. Additionally, we can see that the "duration hours/min" column contains values such as "1/2 hour", "hour" and "1-2 hrs" - there is no clear format. This lack of structure in the data makes it more challenging to work with.

Taking all of the above into consideration, it is clear that we should keep the "duration (seconds)" column and remove the "duration (hours/min)" column. We can remove the unnecessary column via the `DataFrame.drop()` method:



In [58]:
df = df.drop("duration (hours/min)", axis=1)

The column has been removed, which we can double check by looking at all the columns in dataframe (and see that the "durations (hours/min)" column is no longer there):

In [59]:
df.columns

Index(['datetime', 'city', 'state', 'country', 'shape', 'duration (seconds)',
       'comments', 'date posted', 'latitude', 'longitude '],
      dtype='object')

# 7. Saving the processed dataset to a file <a id='saving-the-processed-dataset-to-a-file'></a>

We have now cleaned the data and it is ready to be analyzed. However, before we begin our analysis, let's save the processed version of the dataset, so that we can simply load it in the future, without having to clean the data again.

We can save a `DataFrame` to a csv file via the `DataFrame.to_csv()` method, like so:

In [60]:
df.to_csv("ufo_sightings_clean.csv", sep="\t", index=False)

The first argument (in the `DataFrame.to_csv()` method) is the path to the file. In the above case, it is just the name of the file, since we will be saving it in our current directory. 

The `sep` argument determines what separator should be used when saving the data to a file. Even though **CSV** stands for **comma-separated values**, Pandas allows us to choose a different separator. In the above case, we pass in the string `"\t"` as the separator, which corresponds to the tab character. The tab character is a special character, which is often displayed as a chunk of white space (typically used for indentation). Files which use the tab character for separating values are called **TSV** files. The TSV file format is often preferred over the CSV format, due to the fact that commas frequently occur in text containing natural language. If we save such text in a CSV file, the computer will be unable to determine, which comma separates values and which comma is simply part of some text (and therefore part of a value). The tab character occurs less frequently in text (especially short text) and it often can be cleaned out from the text, without changing the text's meaning or readability.

The `index` argument determines whether or not the index should be saved to the file along with the dataset. Since we have not done anything with the index, saving it is not necessary, as it will only make the file bigger.